### PR TITLE
Improve standard stream handling when buffering

### PR DESCRIPTION
**Pull request by cjwatson**
One of the last remaining issues in porting Launchpad to Python 3 is
that our CI infrastructure runs tests with subunit output (thereby
enabling test output buffering), and a few tests assume that on Python 3
they can use `sys.stdout.buffer` or `sys.stderr.buffer` to get an object
suitable for writing binary data to the standard streams.  These tests
currently fail because when zope.testrunner is doing test output
buffering it replaces the standard streams with instances of
`io.StringIO`.

Now, it's true that https://docs.python.org/3/library/sys.html notes
that "the standard streams may be replaced with file-like objects like
`io.StringIO` which do not support the `buffer` attribute"; however,
there are a couple of cases where it's difficult to work around this,
for example because the assumption is baked into existing libraries like
subunit itself.  Besides, I don't think that this note in the standard
library documentation *necessarily* means that zope.testrunner has to
do something inconvenient.

An alternative approach is to use a variant of `io.TextIOWrapper`, which
provides the `buffer` attribute for tests that need it, and can easily
be made to continue to provide the `getvalue` method for internal use.
This seems like a good compromise.

### CODE REVIEW

**Review by mgedmin — APPROVED** [review]
LGTM!

**Comment by mgedmin** [comment]
Uh, we don't have Travis CI any more, and we haven't set up GitHub Actions for zope.testrunner yet, which makes things a bit awkward.

We have scripts and templates in https://github.com/zopefoundation/meta/tree/master/config that automate setting up GitHub Actions, but I haven't used them much.

**Comment by cjwatson** [comment]
Mm, I've used them a couple of times so I'll see if I can get them working for zope.testrunner later today.

**Comment by mgedmin** [comment]
I've made an attempt (#115, #116), but it's trickier than I thought, and I don't mind if somebody else finishes it without waiting for me.

**Comment by cjwatson** [comment]
I'm starting with the bulk job of cleaning up lint errors at the moment.

**Comment by cjwatson** [comment]
#117 is a bit gigantic, but it makes zope.testrunner `flake8`-clean which should make the rest of the conversion easier.

**Comment by icemac** [comment]
Let's to merge this PR wait until #116 is merged.

**Comment by mgedmin** [comment]
#116 was merged.  Would you mind rebasing this on git master?

**Comment by cjwatson** [comment]
Rebased now.  How's this?

File: src/zope/testrunner/runner.py
Comment (jamadden) [inline src/zope/testrunner/runner.py:916]: `StringIO` is `from six import StringIO`. On Python 2, that's `StringIO.StringIO`, which is the non-native version (`cStringIO.StringIO` is the native version). 

Long story short, I was concerned about possible performance implications, but then I saw the next lines where we were unconditionally assigning `StringIO` instances, so this shouldn't be a regression. Indeed, on Python 3 it might be an improvement!

**Review by jamadden — APPROVED** [review]
LGTM.

File: src/zope/testrunner/runner.py
Comment (cjwatson) [reply src/zope/testrunner/runner.py:916]: Yes, it should be near enough identical on Python 2, and on Python 3 I think `io.StringIO` and an `io.TextIOWrapper` around an `io.BytesIO` are likely to perform similarly enough for our purposes.

I've taken to explicitly using `six.StringIO` or `io.StringIO` in new code rather than doing something like `from six import StringIO`, because the fact that there are multiple kinds of `StringIO` with different semantics is far too confusing (in hindsight it's a shame that the modern version wasn't called `io.TextIO`).  Cleaning that up can be for later though.

**Pull request merged**
